File: java/scon-core/src/main/java/io/github/realmlabs/scon/Analysis.java
package io.github.realmlabs.scon;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public final class Analysis {
    public record SourcePosition(int line, int column) {}
    public record SourceRange(SourcePosition start, SourcePosition end, Span span) {}
    public record Comment(String text, Span span, SourceRange range) {}
    public enum DiagnosticSeverity { Error, Warning, Information, Hint }
    public record Diagnostic(ErrorCode code, String message, DiagnosticSeverity severity, String file, SourceRange range) {}
    public record TokenInfo(String kind, String text, Span span, SourceRange range) {}
    public record Symbol(List<String> path, String file, SourceRange range) {}
    public record Definition(List<String> path, String file, SourceRange range) {}
    public enum ReferenceKind { Substitution, Interpolation, ObjectSpread, ArraySpread }
    public record Reference(List<String> path, ReferenceKind kind, String file, SourceRange range, Definition target) {
        Reference withTarget(Definition target) { return new Reference(path, kind, file, range, target); }
    }
    public record IncludeReference(String path, String file, SourceRange range, String resolvedPath) {}
    public record ParsedSource(String file, List<TokenInfo> tokens, List<Comment> comments, List<Symbol> symbols) {}
    public record DocumentAnalysis(
        String file,
        ParsedSource parsed,
        List<Diagnostic> diagnostics,
        List<Comment> comments,
        List<Symbol> symbols,
        List<Definition> definitions,
        List<Reference> references,
        List<IncludeReference> includes,
        SconValue value
    ) {}

    private Analysis() {}

    public static ParsedSource parseSource(String source, String file) {
        var document = Parser.parseDocument(source, file);
        var lineIndex = new LineIndex(source);
        var tokens = Lexer.lex(source).stream().map(t -> new TokenInfo(t.kind(), t.text(), t.span(), lineIndex.range(t.span()))).toList();
        var comments = comments(tokens);
        return new ParsedSource(file, tokens, comments, symbols(document.root(), lineIndex, file, List.of()));
    }

    public static DocumentAnalysis analyzeSource(String source, String file) {
        var lineIndex = new LineIndex(source);
        List<TokenInfo> tokens = List.of();
        try {
            tokens = Lexer.lex(source).stream().map(t -> new TokenInfo(t.kind(), t.text(), t.span(), lineIndex.range(t.span()))).toList();
            var document = Parser.parseDocument(source, file);
            var comments = comments(tokens);
            var parsed = new ParsedSource(file, tokens, comments, symbols(document.root(), lineIndex, file, List.of()));
            var definitions = definitions(document.root(), lineIndex, file, List.of());
            var references = resolveTargets(references(document.root(), lineIndex, file), definitions);
            var diagnostics = new ArrayList<Diagnostic>();
            SconValue value = null;
            try {
                value = Scon.parseString(source);
            } catch (SconException ex) {
                diagnostics.add(diagnostic(ex, lineIndex, file));
            }
            return new DocumentAnalysis(file, parsed, diagnostics, comments, parsed.symbols(), definitions, references, includes(document.root(), lineIndex, file), value);
        } catch (SconException ex) {
            var comments = comments(tokens);
            return new DocumentAnalysis(file, null, List.of(diagnostic(ex, lineIndex, file)), comments, List.of(), List.of(), List.of(), List.of(), null);
        }
    }

    private static List<Symbol> symbols(Ast.ObjectNode object, LineIndex lineIndex, String file, List<String> prefix) {
        var out = new ArrayList<Symbol>();
        for (var member : object.members()) {
            if (member instanceof Ast.Field field) {
                var path = append(prefix, names(field.path()));
                out.add(new Symbol(path, file, lineIndex.range(field.path().span())));
                if (field.value() instanceof Ast.ObjectValueNode nested) out.addAll(symbols(nested.object(), lineIndex, file, path));
            }
        }
        return out;
    }

    private static List<Definition> definitions(Ast.ObjectNode object, LineIndex lineIndex, String file, List<String> prefix) {
        var out = new ArrayList<Definition>();
        for (var member : object.members()) {
            if (member instanceof Ast.Field field) {
                var path = append(prefix, names(field.path()));
                out.add(new Definition(path, file, lineIndex.range(field.path().span())));
                if (field.value() instanceof Ast.ObjectValueNode nested) out.addAll(definitions(nested.object(), lineIndex, file, path));
            }
        }
        return out;
    }

    private static List<Reference> references(Ast.ObjectNode object, LineIndex lineIndex, String file) {
        var out = new ArrayList<Reference>();
        for (var member : object.members()) {
            if (member instanceof Ast.ObjectSpread spread) out.add(reference(spread.sub().path(), ReferenceKind.ObjectSpread, lineIndex, file));
            if (member instanceof Ast.Field field) out.addAll(valueReferences(field.value(), lineIndex, file));
        }
        return out;
    }

    private static List<Reference> valueReferences(Ast.ValueNode value, LineIndex lineIndex, String file) {
        var out = new ArrayList<Reference>();
        if (value instanceof Ast.SubstitutionNode sub) {
            out.add(reference(sub.path(), ReferenceKind.Substitution, lineIndex, file));
        } else if (value instanceof Ast.StringNode str) {
            for (var part : str.parts()) if (part instanceof Ast.StringInterpolation interpolation) out.add(reference(interpolation.path(), ReferenceKind.Interpolation, lineIndex, file));
        } else if (value instanceof Ast.ArrayNode array) {
            for (var item : array.items()) {
                if (item instanceof Ast.ArraySpread spread) out.add(reference(spread.sub().path(), ReferenceKind.ArraySpread, lineIndex, file));
                if (item instanceof Ast.ArrayValue arrayValue) out.addAll(valueReferences(arrayValue.value(), lineIndex, file));
            }
        } else if (value instanceof Ast.ObjectValueNode object) {
            out.addAll(references(object.object(), lineIndex, file));
        }
        return out;
    }

    private static List<IncludeReference> includes(Ast.ObjectNode object, LineIndex lineIndex, String file) {
        var out = new ArrayList<IncludeReference>();
        for (var member : object.members()) {
            if (member instanceof Ast.Include include) out.add(new IncludeReference(include.path().value(), file, lineIndex.range(include.span()), null));
            if (member instanceof Ast.Field field && field.value() instanceof Ast.ObjectValueNode nested) out.addAll(includes(nested.object(), lineIndex, file));
        }
        return out;
    }

    private static Reference reference(Ast.PathNode path, ReferenceKind kind, LineIndex lineIndex, String file) {
        return new Reference(names(path), kind, file, lineIndex.range(path.span()), null);
    }

    private static List<Reference> resolveTargets(List<Reference> references, List<Definition> definitions) {
        Map<String, Definition> byPath = definitions.stream().collect(Collectors.toMap(d -> String.join("\0", d.path()), d -> d, (a, b) -> a));
        return references.stream().map(r -> r.withTarget(byPath.get(String.join("\0", r.path())))).toList();
    }

    private static List<Comment> comments(List<TokenInfo> tokens) {
        return tokens.stream().filter(token -> token.kind().equals("comment")).map(token -> new Comment(token.text(), token.span(), token.range())).toList();
    }

    private static List<String> names(Ast.PathNode path) {
        return path.segments().stream().map(Ast.PathSegment::value).toList();
    }

    private static List<String> append(List<String> prefix, List<String> suffix) {
        var out = new ArrayList<String>(prefix);
        out.addAll(suffix);
        return out;
    }

    private static Diagnostic diagnostic(SconException ex, LineIndex lineIndex, String file) {
        return new Diagnostic(ex.code(), ex.getMessage(), DiagnosticSeverity.Error, file, ex.span() == null ? null : lineIndex.range(ex.span()));
    }

    private static final class LineIndex {
        private final List<Integer> lines = new ArrayList<>(List.of(0));

        LineIndex(String source) {
            for (int i = 0; i < source.length(); i++) if (source.charAt(i) == '\n') lines.add(i + 1);
        }

        SourceRange range(Span span) {
            return new SourceRange(position(span.start()), position(span.end()), span);
        }

        private SourcePosition position(int offset) {
            int line = 0;
            while (line + 1 < lines.size() && lines.get(line + 1) <= offset) line++;
            return new SourcePosition(line, offset - lines.get(line));
        }
    }
}

File: kotlin/scon-core/src/main/kotlin/io/github/realmlabs/scon/Analysis.kt
package io.github.realmlabs.scon

import java.nio.file.Path

public enum class SconDiagnosticSeverity {
    Error,
    Warning,
    Information,
    Hint,
}

public data class SconDiagnostic(
    val code: SconErrorCode,
    val message: String,
    val severity: SconDiagnosticSeverity,
    val file: Path?,
    val range: SourceRange?,
)

public data class SconSymbol(
    val path: List<String>,
    val file: Path?,
    val range: SourceRange,
)

public enum class SconDefinitionKind {
    Field,
}

public data class SconDefinition(
    val path: List<String>,
    val kind: SconDefinitionKind,
    val file: Path?,
    val range: SourceRange,
)

public enum class SconReferenceKind {
    Substitution,
    Interpolation,
    ObjectSpread,
    ArraySpread,
}

public data class SconReference(
    val path: List<String>,
    val kind: SconReferenceKind,
    val file: Path?,
    val range: SourceRange,
    val target: SconDefinition?,
)

public data class SconIncludeReference(
    val path: String,
    val file: Path?,
    val range: SourceRange,
    val resolvedPath: Path?,
)

public data class SconParsedSource(
    val file: Path?,
    val lineIndex: LineIndex,
    val tokens: List<SconToken>,
    val comments: List<SconComment>,
    val symbols: List<SconSymbol>,
)

public data class SconAnalysis(
    val file: Path?,
    val parsed: SconParsedSource?,
    val diagnostics: List<SconDiagnostic>,
    val comments: List<SconComment>,
    val symbols: List<SconSymbol>,
    val definitions: List<SconDefinition>,
    val references: List<SconReference>,
    val includes: List<SconIncludeReference>,
    val value: SconValue?,
)

public fun parseSourceForAnalysis(
    source: String,
    options: SconParseOptions = SconParseOptions(),
): SconParsedSource {
    val document = parseSource(source, options)
    val lineIndex = LineIndex(source)
    val comments = commentsFromTokens(source, lineIndex, document.tokens)
    val symbols = collectSymbols(document.root, lineIndex, options.sourcePath, emptyList())
    return SconParsedSource(
        file = options.sourcePath,
        lineIndex = lineIndex,
        tokens = document.tokens,
        comments = comments,
        symbols = symbols,
    )
}

public fun analyzeSource(
    source: String,
    parseOptions: SconParseOptions = SconParseOptions(),
    resolveOptions: SconResolveOptions = SconResolveOptions(),
): SconAnalysis {
    val document = try {
        parseSource(source, parseOptions)
    } catch (err: SconException) {
        return parseFailureAnalysis(source, parseOptions.sourcePath, err)
    }
    return analyzeParsedDocument(source, document, parseOptions.sourcePath, resolveOptions)
}

public fun analyzeFile(
    path: Path,
    options: SconResolveOptions = SconResolveOptions(),
): SconAnalysis {
    val source = options.sourceStore.readSource(path)
        ?: return SconAnalysis(
            file = path,
            parsed = null,
            diagnostics = listOf(
                SconDiagnostic(
                    code = SconErrorCode.IncludeNotFound,
                    message = "SCON file not found: $path",
                    severity = SconDiagnosticSeverity.Error,
                    file = path,
                    range = null,
                ),
            ),
            comments = emptyList(),
            symbols = emptyList(),
            definitions = emptyList(),
            references = emptyList(),
            includes = emptyList(),
            value = null,
        )
    return analyzeSource(
        source,
        SconParseOptions(sourceName = path.toString(), sourcePath = path.toAbsolutePath().normalize()),
        options.copy(includeRoot = options.includeRoot ?: path.toAbsolutePath().parent),
    )
}

public fun getPath(value: SconValue, path: String): SconValue {
    var current = value
    val segments = parsePathQuery(path)
    for (segment in segments) {
        current = (current as? SconValue.ObjectValue)?.values?.get(segment)
            ?: throw SconException(
                SconError(
                    code = SconErrorCode.MissingReference,
                    message = "path ${segments.joinToString(".")} is not defined",
                ),
            )
    }
    return current
}

public fun diagnosticFromError(error: SconError, source: String, file: Path? = null): SconDiagnostic {
    val lineIndex = LineIndex(source)
    return SconDiagnostic(
        code = error.code,
        message = error.message,
        severity = SconDiagnosticSeverity.Error,
        file = file,
        range = error.span?.let(lineIndex::sourceRange),
    )
}

private fun parseFailureAnalysis(source: String, file: Path?, err: SconException): SconAnalysis {
    val lineIndex = LineIndex(source)
    val comments = runCatching {
        commentsFromTokens(source, lineIndex, Lexer(source, file?.toString() ?: "<source>").lex())
    }.getOrElse { emptyList() }
    return SconAnalysis(
        file = file,
        parsed = null,
        diagnostics = listOf(diagnosticFromError(err.error, source, file)),
        comments = comments,
        symbols = emptyList(),
        definitions = emptyList(),
        references = emptyList(),
        includes = emptyList(),
        value = null,
    )
}

private fun analyzeParsedDocument(
    source: String,
    document: ParsedDocument,
    file: Path?,
    resolveOptions: SconResolveOptions,
): SconAnalysis {
    val lineIndex = LineIndex(source)
    val comments = commentsFromTokens(source, lineIndex, document.tokens)
    val semantic = SemanticCollector(lineIndex, file)
    semantic.collectObject(document.root, emptyList())
    val symbols = collectSymbols(document.root, lineIndex, file, emptyList())
    val parsed = SconParsedSource(
        file = file,
        lineIndex = lineIndex,
        tokens = document.tokens,
        comments = comments,
        symbols = symbols,
    )
    val (diagnostics, value) = try {
        emptyList<SconDiagnostic>() to resolveDocument(document, resolveOptions)
    } catch (err: SconException) {
        listOf(diagnosticFromError(err.error, source, file)) to null
    }
    return SconAnalysis(
        file = file,
        parsed = parsed,
        diagnostics = diagnostics,
        comments = comments,
        symbols = symbols,
        definitions = semantic.definitions,
        references = semantic.references,
        includes = semantic.includes,
        value = value,
    )
}

private fun collectSymbols(
    obj: AstObject,
    lineIndex: LineIndex,
    file: Path?,
    prefix: List<String>,
): List<SconSymbol> {
    val out = mutableListOf<SconSymbol>()
    for (member in obj.members) {
        val field = member as? AstField ?: continue
        val path = prefix + field.path.segments.map { it.value }
        out += SconSymbol(path, file, lineIndex.sourceRange(field.path.span))
        val objectValue = field.value as? AstObjectValue
        if (objectValue != null) {
            out += collectSymbols(objectValue.value, lineIndex, file, path)
        }
    }
    return out
}

private class SemanticCollector(
    private val lineIndex: LineIndex,
    private val file: Path?,
) {
    private val completed = mutableMapOf<List<String>, SconDefinition>()
    val definitions = mutableListOf<SconDefinition>()
    val references = mutableListOf<SconReference>()
    val includes = mutableListOf<SconIncludeReference>()

    fun collectObject(obj: AstObject, prefix: List<String>) {
        for (member in obj.members) {
            when (member) {
                is AstObjectSpread -> pushReference(
                    member.substitution.path,
                    SconReferenceKind.ObjectSpread,
                    member.substitution.path.span,
                )
                is AstInclude -> includes += SconIncludeReference(
                    path = member.path.value,
                    file = file,
                    range = lineIndex.sourceRange(member.path.span),
                    resolvedPath = resolveIncludePath(file, member.path.value),
                )
                is AstField -> {
                    val path = prefix + member.path.segments.map { it.value }
                    val objectValue = member.value as? AstObjectValue
                    if (objectValue != null) {
                        pushDefinition(path, member.path.span)
                        collectObject(objectValue.value, path)
                    } else {
                        collectValue(member.value)
                        pushDefinition(path, member.path.span)
                    }
                }
            }
        }
    }

    private fun collectValue(value: AstValue) {
        when (value) {
            is AstObjectValue -> collectObject(value.value, emptyList())
            is AstArray -> value.items.forEach {
                when (it) {
                    is AstArrayValueItem -> collectValue(it.value)
                    is AstArraySpread -> pushReference(
                        it.substitution.path,
                        SconReferenceKind.ArraySpread,
                        it.substitution.path.span,
                    )
                }
            }
            is AstString -> value.parts.forEach {
                if (it is AstStringInterpolationPart) {
                    pushReference(it.path, SconReferenceKind.Interpolation, it.path.span)
                }
            }
            is AstSubstitution -> pushReference(value.path, SconReferenceKind.Substitution, value.path.span)
            is AstBool, is AstNull, is AstNumber -> Unit
        }
    }

    private fun pushDefinition(path: List<String>, span: SourceSpan) {
        val definition = SconDefinition(path, SconDefinitionKind.Field, file, lineIndex.sourceRange(span))
        completed[path] = definition
        definitions += definition
    }

    private fun pushReference(path: AstPath, kind: SconReferenceKind, span: SourceSpan) {
        val names = path.segments.map { it.value }
        references += SconReference(names, kind, file, lineIndex.sourceRange(span), completed[names])
    }
}

private fun resolveIncludePath(file: Path?, includePath: String): Path? =
    runCatching {
        validateIncludePath(includePath)
        (file?.parent ?: Path.of(".")).resolve(includePath).normalize()
    }.getOrNull()

private fun validateIncludePath(includePath: String) {
    if (
        includePath.contains("://") ||
        includePath.startsWith("classpath:") ||
        includePath.contains('*') ||
        includePath.startsWith('~') ||
        includePath.startsWith('$') ||
        includePath.startsWith('/') ||
        looksLikeWindowsAbsolute(includePath)
    ) {
        throw IllegalArgumentException("invalid include path")
    }
}

private fun looksLikeWindowsAbsolute(path: String): Boolean =
    path.length >= 3 && path[1] == ':' && (path[2] == '/' || path[2] == '\\')

private fun parsePathQuery(path: String): List<String> {
    val parts = mutableListOf<String>()
    val current = StringBuilder()
    var index = 0
    while (index < path.length) {
        when (val ch = path[index]) {
            '.' -> {
                if (current.isEmpty()) {
                    throw SconException(SconError(SconErrorCode.UnexpectedToken, "empty path segment"))
                }
                parts += current.toString()
                current.clear()
            }
            '"' -> {
                index += 1
                while (index < path.length && path[index] != '"') {
                    if (path[index] == '\\' && index + 1 < path.length) index += 1
                    current.append(path[index])
                    index += 1
                }
            }
            else -> current.append(ch)
        }
        index += 1
    }
    if (current.isNotEmpty()) parts += current.toString()
    if (parts.isEmpty()) throw SconException(SconError(SconErrorCode.UnexpectedToken, "empty path"))
    return parts
}
